File: Data Structures & Algorithms/house-robber-ii/submission-3.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        cache = [-1] * len(nums)
        
        return max(self.dfs(0, nums[:-1], [-1] * len(nums)), self.dfs(0, nums[1:], [-1] * len(nums)))
    
    def dfs(self, i, nums, cache):
        if i >= len(nums):
            return 0
        if cache[i] != -1:
            return cache[i]
        cache[i] = max(nums[i] + self.dfs(i+2, nums, cache), self.dfs(i+1, nums, cache))
        return cache[i]
    
# [2,9,8,3,6]

File: Data Structures & Algorithms/jump-game/submission-3.java
class Solution {
    public boolean canJump(int[] nums) {
        int n = nums.length;
        boolean[] dp = new boolean[n];
        dp[n - 1] = true;

        for(int i = n - 2; i >= 0; --i) {
            for(int j = i + 1; j <= i + nums[i]; ++j) {
                if(j < n && dp[j]) {
                    dp[i] = true;
                }
            }
        }

        return dp[0];
    }
}
